config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurer.java

```diff
@@ -301,14 +314,26 @@ public OpaqueTokenConfigurer introspectionUri(String introspectionUri) {
 		public OpaqueTokenConfigurer introspectionClientCredentials(String clientId, String clientSecret) {
 			Assert.notNull(clientId, "clientId cannot be null");
 			Assert.notNull(clientSecret, "clientSecret cannot be null");
-			this.introspectionClientId = clientId;
-			this.introspectionClientSecret = clientSecret;
+			RestTemplate restTemplate = new RestTemplate();
+			restTemplate.getInterceptors().add(new BasicAuthenticationInterceptor(clientId, clientSecret));
+			this.restOperations = restTemplate;
 			return this;
 		}
 
-		AuthenticationProvider getProvider() {
-			return new OAuth2IntrospectionAuthenticationProvider(this.introspectionUri,
-					this.introspectionClientId, this.introspectionClientSecret);
+		public OpaqueTokenConfigurer introspectionClient(OAuth2TokenIntrospectionClient introspectionClient) {
+			Assert.notNull(introspectionClient, "introspectionClient cannot be null");
+			this.introspectionClient = introspectionClient;
+			return this;
+		}
+
+		OAuth2TokenIntrospectionClient getIntrospectionClient() {
+			if (this.introspectionClient != null) {
+				return this.introspectionClient;
+			}
+			if (this.introspectionUri != null || this.restOperations != null) {
```

Comment:
The operator should be `&&` not `||`

Reply:
This is by design to guard against partial configuration. Changing to `&&` could lead partial configuration to prioritize the wrong thing (instead of just failing as it should).

For example:

```java
http
    .oauth2ResourceServer()
        .opaqueToken()
            .introspectionUri(uri)
```

This is only partially configured and is erroneous. However, if a `@Bean` were also published, then using `&&` would end up publishing the bean instead of failing on partial configuration.

That said, I think this logic reads in a confusing way, so I will express it differently. I think this is also an opportunity to make sure `.introspectionClient(client)` and `introspectionUri(uri).introspectionClientDetails(id, secret)` get equal priority, like `jwtDecoder` and `jwkSetUri` do.